packages/backend-common/src/database/types.ts

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import knex from 'knex';
+import { Config } from '@backstage/config';
+
+/**
+ * DatabaseConfiguration exposes the Application configuration to PluginDatabaseManager instances.
+ */
+export interface DatabaseConfiguration {
+  /**
+   * Returns the database configuration for the given Plugin.
+   */
+  getDatabaseConfig(pluginId: string): Config;
+
+  /**
+   * Returns the database configuration for administrating the database server.
+   */
+  getAdminConfig(): Config;
+}
+
+/**
+ * The PluginDatabaseManager produces PluginDatabaseFactory objects for plugins to consume and store their data.
+ */
+export type PluginDatabaseManager = {
+  /**
+   * getDatabaseFactory returns the database factory object for plugins to obtain database connections.
+   */
+  getDatabaseFactory(pluginId: string): PluginDatabaseFactory;
+};
+
+/**
+ * The PluginDatabaseFactory is used to provide a mechanism for backend plugins to obtain database connections for
+ * itself.
+ *
+ * The purpose of this factory is to allow plugins to get isolated data stores so that plugins are discouraged from database integration.
+ */
+export type PluginDatabaseFactory = {
+  /**
+   * Returns a database connection. Plugins can omit the `database` parameter to get the default plugin database, or
+   * provide an identifier that will be used to identify a separate database from the default.
```

Comment:
@andrewthauer suggested to remove it because it's probably overkill for what we have, so it's no longer in this PR. I don't have a need for it now, but there could be application-level sharding, personally identifiable information sharding, etc.

Reply:
In general I think we want to encourage backend plugins to be quite lean, and something like PII is likely better off in a separate system. Definitely something to keep out of scope for now either way. I think forwards compatibility is pretty straightforward too right? Basically just appending `_${id}` if it's not the default database.